test/system/src/test/java/io/pravega/test/system/MultiReaderWriterTxnWithFailoverTest.java

```diff
@@ -0,0 +1,451 @@
+
+        //Scale down controller instances to 2
+        controllerInstance.scaleService(2, true);
+        //zookeeper will take about 60 seconds to detect that the node has gone down
+        Thread.sleep(60000);
+        log.info("Scaling down controller instances from 3 to 2");
+
+        currentWriteCount2 = eventWriteCount.get();
+        currentReadCount2 = eventReadCount.get();
+
+        log.info("Read count: {}, write count: {} after controller failover after sleep", currentReadCount2, currentWriteCount2);
+
+        //ensure writes are happening
+        assertTrue(currentWriteCount2 > currentWriteCount1);
+        //ensure reads are happening
+        assertTrue(currentReadCount2 > currentReadCount1);
+
+        //Scale down SSS, controller to 1 instance each.
+        segmentStoreInstance.scaleService(1, true);
+        controllerInstance.scaleService(1, true);
+        //zookeeper will take about 60 seconds to detect that the node has gone down
+        Thread.sleep(new Random().nextInt(50000) + 60000);
+        log.info("Scaling down  to 1 controller, 1 SSS instance");
+
+        currentWriteCount1 = eventWriteCount.get();
+        currentReadCount1 = eventReadCount.get();
+
+        log.info("Stop write flag status: {}, stop read flag status: {} ", stopWriteFlag, stopReadFlag);
+        log.info("Event data: {}, read count: {}", eventData.get(), eventReadCount.get());
+        log.info("Read count: {}, write count: {} with SSS and controller failover after sleep", currentReadCount1, currentReadCount1);
+    }
+
+    private CompletableFuture<Void> startWritingIntoTxn(final AtomicLong data, final EventStreamWriter<Long> writer,
+                                                 final AtomicBoolean stopWriteFlag, final AtomicLong eventWriteCount) {
+        return CompletableFuture.runAsync(() -> {
+            while (!stopWriteFlag.get()) {
+                Transaction<Long> transaction = null;
+                    try {
+                        transaction = retry
+                                .retryingOn(MultiReaderWriterTxnWithFailoverTest.TxnCreationFailedException.class)
+                                .throwingOn(RuntimeException.class)
+                                .run(() -> createTransaction(writer, stopWriteFlag));
+
+                        //each transaction has 50 events
+                        for (int j = 0; j < 50; j++) {
+                            long value = data.incrementAndGet();
+                            Thread.sleep(100);
+                            transaction.writeEvent(String.valueOf(value), value);
+                            log.debug("Writing event: {} into transaction: {}", value, transaction);
+                        }
+                        //commit Txn
+                        transaction.commit();
+
+                        //wait for transaction to get committed
+                        while (!transaction.checkStatus().equals(Transaction.Status.COMMITTED)) {
```

Comment:
also you may use Retry.runAsync so that you are not blocking a thread to verify txn status. 
(this way you will allow for multiple txns to run concurrently while readers are reading data from them in the background).